src/shared/fileRegistry.ts

```diff
@@ -0,0 +1,205 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { getLogger } from './logger/logger'
+import * as pathutils from './utilities/pathUtils'
+import * as path from 'path'
+
+export interface WorkspaceItem<T> {
+    path: string
+    item: T
+}
+
+/**
+ * WorkspaceFileRegistry lets us index files in the current registry. It is used
+ * for CFN templates among other things
+ */
+export abstract class WorkspaceFileRegistry<T> implements vscode.Disposable {
```

Comment:
Optional naming nits (since we are changing the name anyway; in general we should avoid ornate names unless there is a clear need for a name that differs from some other existing, non-ornate name):

- "registry" is vague yet ornate. 
- in theory we might track files outside of the VSCode Workspace?
 
Suggest `WatchedFiles` (each file is a `WatchedFile`, singular). Or perhaps `TrackedFiles` + `TrackedFile`.

```suggestion
export abstract class WatchedFiles<T> implements vscode.Disposable {
```

Reply:
Sounds good to me